Fix: scrubber

3lvis opened the pull request:
Fixes a few issues with the scrubber calculating it's position. This works with all combinations of hidden or shown status bar, navigation bar, tab bar, and so on.

3lvis commented:
Right now there's a bug that I haven't been able to inspect closely.

marijnschilling commented on Sources/SectionScrubber.swift:
```diff
@@ -109,30 +107,30 @@ public class SectionScrubber: UIView {
-        self.scrubberGestureView.addGestureRecognizer(self.longPressGestureRecognizer)
-        self.scrubberGestureView.addGestureRecognizer(self.dragGestureRecognizer)
-        self.addSubview(scrubberGestureView)
+        self.addGestureRecognizer(self.longPressGestureRecognizer)
```
I added a scrubberGestureView because I find it strange behaviour if you can scrub on the left side of the scrubberView. It makes more sense if you can only scrub when you're pressing the scrubber thingy itself. But because this is just a bit to small to press comfortably I added this view of 44\* width to be able to press the scrubber more intuitively.

I copied this behaviour from the google photo's app.
- from HIG: https://developer.apple.com/ios/human-interface-guidelines/visual-design/layout/

3lvis replied:
Makes total sense. Fixing that.

marijnschilling commented on Sources/SectionScrubber.swift:
```diff
@@ -18,6 +20,14 @@ public class SectionScrubber: UIView {
 
+    /*
+     When calculating the NSIndexPath for the current scrubber position we need to use a x and a y coordinate,
+     the y coordinate is provided by how far you have scrubber the scrubber, meanwhile we use a hardcoded x since
```
small typo: scrubbed the scrubber

marijnschilling commented on Sources/SectionScrubber.swift:
```diff
@@ -18,6 +20,14 @@ public class SectionScrubber: UIView {
+     using 5 would ensure us that most of the time the first item in each row will be selected to retreive the
+     index path at certain location.
+    */
+    static let initialXCoordinateToCalculateIndexPath = CGFloat(5)
```
Would the middle of the screen be safest? 🤔

3lvis replied:
@marijnschilling middle of the screen might make it think there are no cells in a row like this. Any way I can make the comment more clear to describe that? That's what I was trying to explain.

![group](https://cloud.githubusercontent.com/assets/1088217/16734343/7d627014-4785-11e6-9fa4-633ed0055f49.png)

3lvis replied:
👀

marijnschilling commented on iOS/RemoteCollectionController.swift:
```diff
@@ -4,7 +4,8 @@ class RemoteCollectionController: UICollectionViewController {
 
     lazy var overlayView: UIView = {
-        let view = UIView()
+        // Needs a frame, otherwise it won't show up.
```
Might be a bit obvious comment right? :joy: Or am I missing something?

3lvis replied:
It gets a frame here, so it shouldn't be needed but UIKit = :dolan:

https://github.com/bakkenbaeck/SectionScrubber/blob/1a48e4e39e5053a778254292211e3f8bfe3afc70/iOS/RemoteCollectionController.swift#L74

3lvis replied:
Oh, lol. Now I see the problem.

marijnschilling commented:
Ir would be nice if you could also fix the tests. I found it really useful for debugging to have it clear and separated how you get the yOffset for the sectionScrubber form the translation of the gestureRecognizer. And the same for how you get the position for the SectionScrubber from the collectionView.contentOffset.y.

Or just remove the tests all together. Now they fail like a snail 🐌  👎 

And it would be nice if you could update the Readme 😬

3lvis commented:
@marijnschilling Noted 👯

marijnschilling commented:
For the rest I think it looks really good and much much clearer! 

The only thing is that the `handleScrub` and `updateScrubberPosition` could be a more transparant. I think if people will use this element and have problems finding the right offset for their collectionView it could be useful to make these function really super clear. Writing tests will probably force these functions to split up a bit and make it more general as well.

3lvis commented:
> I think if people will use this element and have problems finding the right offset for their collectionView

What do you mean?

3lvis commented:
Removed the tests and added a separate issue for it.

I think it could be great if we moved away from UICollectionView again and just injected the dependencies from the Scrubber as you did in the beginning, @marijnschilling.

marijnschilling commented:
With my comment I just meant that if people using this element have issues with the scrubber not behaving correctly (i.e. scrolling too far down/not far enough etc.) It might help to make these methods (handleScrub and updateScrubberPosition) really transparant. 

Now it works for us, with the tabbar and te navigationbar, but you never know in what strange viewHierarchy other people will wan tot use the scrubber and might need to make small adjustments. 

I just thought these methods need to be really clear/transparant, broken up in small pieces etc. to make adjustments as easy as possible.  But I think writing tests for these function will force you to break the functions themselves up in smaller, clearer pieces.  

So let's look into this in #31 👏

marijnschilling commented:
That bug that it moves in opposite direction when you let go of the scrubber is sooo weird btw! I looked into it a bit, but still no clue. 🤔

marijnschilling commented on Sources/SectionScrubber.swift:
```diff
@@ -121,13 +121,13 @@ public class SectionScrubber: UIView {
         self.longPressGestureRecognizer.cancelsTouchesInView = false
         self.longPressGestureRecognizer.delegate = self
-        self.addGestureRecognizer(self.longPressGestureRecognizer)
+        self.scrubberImageView.addGestureRecognizer(self.longPressGestureRecognizer)
```
I see you've added the gesture recogniser to the scrubberImageView, but this imageView is only 26x56 and because of the small width still hard to touch sometimes. I noticed this behaviour when I was working on it earlier and thats why I added an _extra_ special gesture-view of 56x44. I worked really well!

3lvis commented:
https://github.com/bakkenbaeck/SectionScrubber/issues/32

3lvis commented:
https://github.com/bakkenbaeck/SectionScrubber/issues/33

3lvis commented:
Added issues for the remaining things in this PR.

marijnschilling commented:
Looks good! Merge it :v:
